feat(timeline): add timeline API

Oupsla opened the pull request:
Follow-up : https://github.com/MadKudu/node-hubspot/pull/90
Resolving : #81


@MattMSumner here is the pull request with only the timeline API
I can't make the test work but I'm close to a solution I guess with the nock library (if you can look at it)

Oupsla committed:
feat(timeline): add timeline API

MattMSumner commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "index.js",
   "scripts": {
-    "test": "npm run lint && npm run mocha && npm run tsc",
+    "test": "NODE_ENV=test npm run lint && npm run mocha && npm run tsc",
```
👍 Thanks for adding this

MattMSumner commented on test/timeline.js:
```diff
@@ -0,0 +1,90 @@
+describe('timeline', function() {
+  let hubspot
+
+  before(() => {
```
I think this is a step in the right direction. Instead of refreshing the token we could write these tests on the assumption that you already have a un-expired token. I was thinking something along the lines of:

```js
  before(() => {
    nockHelper.mockOauthEndpoint(
      /\/integrations\/v1\/[0-9]+\/timeline\/event-types'/,
      {
        id: 261,
        name: 'Test Event Type',
        headerTemplate:
          '# Title for event {{id}}\nThis is an event for {{objectType}}',
        detailTemplate:
          'This event happened on {{#formatDate timestamp}}{{/formatDate}}',
        applicationId: 123,
        objectType: 'CONTACT',
      },
    )

    hubspot = new Hubspot({
      accessToken: process.env.ACCESS_TOKEN || 'some-fake-token',
    })
  })
  after(nockHelper.resetNock)
```

You could then hit the real api in the test with an access token with something like:

```
NOCK_OFF=true ACCESS_TOKEN=my-super-long-and-real-access-token npm run mocha -- --grep "broadcasts"
```

MattMSumner replied:
For this to work you would need something like the following in nockHelper:

```js
// test/nock_helper.js
const nock = require('nock')

class NockHelper {
  mockRateLimit() {
    nock('http://api.hubapi.com', { encodedQueryParams: true })
      .get('/integrations/v1/limit/daily')
      .query({ hapikey: 'demo' })
      .reply(200, [
        {
          name: 'api-calls-daily',
          usageLimit: 160000,
          currentUsage: 19742,
          collectedAt: Date.now(),
          fetchStatus: 'CACHED',
          resetsAt: Date.now() + 24 * 60 * 60 * 1000,
        },
      ])
  }

  mockEndpoint(path, data) {
    return () => {
      nock.disableNetConnect()
      this.mockRateLimit()
      nock('http://api.hubapi.com', { encodedQueryParams: true })
        .get(path)
        .query({ hapikey: 'demo' })
        .reply(200, data)
    }
  }

  mockOauthEndpoint(path, data) {
    return () => {
      nock.disableNetConnect()
      nock('http://api.hubapi.com')
        .get(path)
        .reply(200, data)
    }
  }

  resetNock() {
    nock.cleanAll()
    nock.enableNetConnect()
  }
}

module.exports = new NockHelper()

MattMSumner replied:
@Oupsla I've just pushed a PR up with an example test using oauth and nock https://github.com/MadKudu/node-hubspot/pull/131

MattMSumner commented on lib/timeline.js:
```diff
@@ -0,0 +1,107 @@
+  }
+}
+
+function guid() {
```
How would you feel about moving this to it's own file and importing it? Maybe something like `lib/utils/guid.js`?

MattMSumner commented on lib/timeline.js:
```diff
@@ -0,0 +1,107 @@
+  updateEventTypeProperty(eventId, applicationId, propertyId, data, cb) {
+    const parameters = {
+      method: 'PUT',
+      path: `/integrations/v1/${applicationId}/timeline/event-types/${eventId}/properties/${propertyId}`,
```
Looks like this endpoint is not consistent with other hubspot endpoints. It seems to be a PUT to `/integrations/v1/${applicationId}/timeline/event-types/${eventId}/properties` and the id needs to be in the body of the request.

You could do something like:

```js
updateEventTypeProperty(eventId, applicationId, propertyId, data, cb) {
  data['id'] = data['id'] || propertyId
  const parameters = {
    method: 'PUT',
    path: `/integrations/v1/${applicationId}/timeline/event-types/${eventTypeId}/properties`,
    body: data,
  }

  return this.client._request(parameters, cb)
}
```

https://developers.hubspot.com/docs/methods/timeline/udpate-timeline-event-type-property

MattMSumner commented on lib/timeline.js:
```diff
@@ -0,0 +1,107 @@
+
+  createTimelineEvent(applicationId, eventTypeId, data, cb) {
+    if (
+      !this.client.clientId ||
```
I'm not sure this is true. I've been able to use the Timeline API with only an oauth token. So if we wanted to raise an error we could do:

```js
if (!this.client.auth) {
  return Promise.reject(new Error(...))
}
```
But I think we should be ok not making this check here.

MattMSumner commented on lib/timeline.js:
```diff
@@ -0,0 +1,107 @@
+    if (!data || (!data.email && !data.utk && !data.objectId)) {
+      return cb(
+        new Error(
+          'You must specify at least one of those in data : objectId, email, utk',
```
```suggestion
          'You must specify at least one of the following in data : objectId, email, utk',
```

MattMSumner commented on lib/timeline.js:
```diff
@@ -0,0 +1,107 @@
+      )
+    }
+
+    if (!data || (!data.email && !data.utk && !data.objectId)) {
```
I've tested not passing these and the HubSpot API returns an appropriate error if one of these values are not present. I think it's fine for us not to make this check.

MattMSumner committed:
Add test coverage for timeline events (#1)

Why:

We would like to have some tests around timeline events to avoid
breaking changes.

This PR:

* Introduce dotenv for running tests
* Adds test cases for timeline
* Adds `id` in updateEventTypeProperty and corrects the url
* Removes checks already covered by hubspot
* Pluralize timeline
